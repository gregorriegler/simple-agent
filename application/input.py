from typing import Callable, List, Optional


class Input:

    def __init__(self, display, escape_detector: Optional[Callable[[], bool]] = None):
        self.display = display
        self._stack: List[str] = []
        self._escape_detector = escape_detector

    def stack(self, message: str):
        self._stack.append(message)

    def has_stacked_messages(self) -> bool:
        return bool(self._stack)

    def read(self) -> str:
        if self._stack:
            return self._stack.pop()
        return self.display.input()

    def escape_requested(self) -> bool:
        if self._escape_detector is None:
            return False
        return self._escape_detector()

    def set_escape_detector(self, escape_detector: Callable[[], bool]):
        self._escape_detector = escape_detector
