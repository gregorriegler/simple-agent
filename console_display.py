from agent import Display
import sys


class ConsoleDisplay(Display):

    def __init__(self, indent_level=0, print_fn=print, agent_type="Agent"):
        self.indent_level = indent_level
        self.print = print_fn
        self.agent_type = agent_type
        self.base_indent = "       " * (indent_level + 1)  # 7 spaces per level
        self.agent_prefix = "       " * indent_level + f"{agent_type}: "

    def assistant_says(self, message):
        lines = str(message).split('\n')
        if lines:

            self.print(f"\n{self.agent_prefix}{lines[0]}")
            for line in lines[1:]:
                self.print(f"{self.base_indent}{line}")

    def tool_about_to_execute(self, parsed_tool):
        indented_tool = self._indent_lines(str(parsed_tool), "       " * self.indent_level)
        self.print(f"\n{indented_tool}")

    def tool_result(self, result):
        result = self._indent_lines(result, self.base_indent)
        self.print(f"\n{result}")

    def input(self):
        prompt = "       " * self.indent_level + "Press Enter to continue or type a message to add: "
        self.print(f"\n{prompt}", file=sys.stderr)
        return input("")

    def exit(self):
        exit_msg = "       " * self.indent_level + "Exiting..."
        self.print(f"\n{exit_msg}")

    def _indent_lines(self, text, prefix="    "):
        lines = str(text).split('\n')
        return '\n'.join(f"{prefix}{line}" if line.strip() else line for line in lines)
