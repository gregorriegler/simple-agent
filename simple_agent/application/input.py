from typing import List

from .user_input import UserInput


class Input:

    def __init__(self, user_input: UserInput):
        self.user_input = user_input
        self._stack: List[str] = []

    def stack(self, message: str):
        self._stack.append(message)

    def has_stacked_messages(self) -> bool:
        return bool(self._stack)

    def read(self) -> str:
        if self._stack:
            return self._stack.pop()
        return self.user_input.read()

    def escape_requested(self) -> bool:
        return self.user_input.escape_requested()
