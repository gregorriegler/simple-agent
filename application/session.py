from dataclasses import dataclass

from application.agent import Agent
from application.llm import Messages
from application.persisted_messages import PersistedMessages
from application.session_storage import SessionStorage
from application.system_prompt import SystemPrompt
from tools import ToolLibrary


@dataclass
class SessionArgs:
    continue_session: bool
    start_message: str | None


def run_session(
    continue_session: bool,
    user_input,
    display,
    session_storage: SessionStorage,
    chat,
    system_prompt: SystemPrompt
):
    messages = session_storage.load() if continue_session else Messages()
    persisted_messages = PersistedMessages(messages, session_storage)

    if continue_session:
        display.continue_session()
    else:
        display.start_new_session()
    tool_library = ToolLibrary(chat)
    system_prompt = system_prompt()
    agent = Agent(chat, system_prompt, user_input, tool_library, display, session_storage)
    agent.start(persisted_messages)
