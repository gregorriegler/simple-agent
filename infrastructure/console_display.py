from application.display import Display
import sys


class ConsoleDisplay(Display):

    def __init__(self, indent_level=0, agent_name="Agent", input_fn=input, print_fn=print):
        self.indent_level = indent_level
        self.print = print_fn
        self.base_indent = "       " * (indent_level + 1)  # 7 spaces per level
        self.agent_prefix = "       " * indent_level + f"{agent_name}: "
        self.input_fn = input_fn

    def assistant_says(self, message):
        lines = str(message).split('\n')
        if lines:

            self.print(f"\n{self.agent_prefix}{lines[0]}")
            for line in lines[1:]:
                self.print(f"{self.base_indent}{line}")

    def tool_result(self, result):
        lines = str(result).split('\n')
        first_three_lines = '\n'.join(lines[:3])
        if len(lines) > 3:
            first_three_lines += '\n... (truncated)'
        result = self._indent_lines(first_three_lines, self.base_indent)
        self.print(f"\n{result}")

    def continue_session(self):
        self.print("Continuing session")

    def start_new_session(self):
        self.print("Starting new session")

    def input(self) -> str:
        prompt = "       " * self.indent_level + "Press Enter to continue or type a message to add: "
        self.print(f"\n{prompt}", file=sys.stderr)
        return self.input_fn("").strip()

    def exit(self):
        exit_msg = "       " * self.indent_level + "Exiting..."
        self.print(f"\n{exit_msg}")

    def _indent_lines(self, text, prefix="    "):
        lines = str(text).split('\n')
        return '\n'.join(f"{prefix}{line}" if line.strip() else line for line in lines)
