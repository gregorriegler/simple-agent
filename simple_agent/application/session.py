from dataclasses import dataclass

from simple_agent.application.agent import Agent
from simple_agent.application.llm import Messages
from simple_agent.application.persisted_messages import PersistedMessages
from simple_agent.application.session_storage import SessionStorage
from simple_agent.application.system_prompt import SystemPrompt
from simple_agent.tools import ToolLibrary


@dataclass
class SessionArgs:
    continue_session: bool
    start_message: str | None
    show_system_prompt: bool = False


def run_session(
    continue_session: bool,
    user_input,
    display,
    session_storage: SessionStorage,
    chat,
    system_prompt: SystemPrompt
):
    messages = session_storage.load() if continue_session else Messages()
    persisted_messages = PersistedMessages(messages, session_storage)

    if continue_session:
        display.continue_session()
    else:
        display.start_new_session()
    tool_library = ToolLibrary(chat)
    system_prompt = system_prompt()
    agent = Agent(chat, system_prompt, user_input, tool_library, display, session_storage)
    agent.start(persisted_messages)
