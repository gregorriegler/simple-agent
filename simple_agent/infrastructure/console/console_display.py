import os
import sys

from simple_agent.application.display import Display
from simple_agent.application.io import IO
from simple_agent.application.tool_library import ToolResult
from simple_agent.infrastructure.stdio import StdIO


def _indent_lines(text, prefix="    "):
    lines = str(text).split("\n")
    return "\n".join(f"{prefix}{line}" if line.strip() else line for line in lines)


class ConsoleDisplay(Display):

    def __init__(self, indent_level, agent_name, io):
        self.indent_level = indent_level
        self.io = io
        self.base_indent = "       " * (indent_level + 1)
        self.agent_prefix = "       " * indent_level + f"{agent_name}: "

    def assistant_says(self, message):
        lines = str(message).split("\n")
        if lines:
            self.io.print(f"\n{self.agent_prefix}{lines[0]}")
            for line in lines[1:]:
                self.io.print(f"{self.base_indent}{line}")

    def user_says(self, message):
        pass

    def tool_call(self, call_id, tool):
        pass

    def tool_result(self, call_id, result: ToolResult):
        message = result.message if result else ""
        if result and result.display_language == "diff" and result.display_body:
            colored_diff = _colorize_diff_for_console(result.display_body)
            if "\n\n" in message:
                summary, _, rest = message.partition("\n\n")
                message = f"{summary}\n\n{colored_diff}"
            else:
                message = colored_diff

        lines = message.split("\n") if message else [""]
        if result.success:
            display_lines = lines[:3]
            if len(lines) > 3:
                display_lines = display_lines + ['... (truncated)']
        else:
            display_lines = lines
        display_text = "\n".join(display_lines)
        formatted = _indent_lines(display_text, self.base_indent)
        self.io.print(f"\n{formatted}")

    def continue_session(self):
        self.io.print("Continuing session")

    def start_new_session(self):
        self.io.print("Starting new session")

    def waiting_for_input(self):
        pass

    def interrupted(self):
        interrupted_msg = "       " * self.indent_level + "[Session interrupted by user]"
        self.io.print(f"\n{interrupted_msg}")

    def exit(self):
        exit_msg = "       " * self.indent_level + "Exiting..."
        self.io.print(f"\n{exit_msg}")


def _supports_color() -> bool:
    if os.environ.get("NO_COLOR"):
        return False
    stream = sys.stdout
    return bool(stream.isatty())


def _colorize_diff_for_console(diff_text: str) -> str:
    if not diff_text or not _supports_color():
        return diff_text

    reset = "\033[0m"
    colors = {
        "header_old": "\033[31;1m",  # bold red
        "header_new": "\033[32;1m",  # bold green
        "hunk": "\033[36m",          # cyan
        "removed": "\033[31m",       # red
        "added": "\033[32m",         # green
        "info": "\033[90m",          # bright black
    }

    colored_lines = []
    for line in diff_text.split("\n"):
        color = ""
        if line.startswith("---"):
            color = colors["header_old"]
        elif line.startswith("+++"):
            color = colors["header_new"]
        elif line.startswith("@@"):
            color = colors["hunk"]
        elif line.startswith("+"):
            color = colors["added"]
        elif line.startswith("-"):
            color = colors["removed"]
        elif line.startswith("\\ No newline"):
            color = colors["info"]

        if color:
            colored_lines.append(f"{color}{line}{reset}")
        else:
            colored_lines.append(line)

    return "\n".join(colored_lines)
