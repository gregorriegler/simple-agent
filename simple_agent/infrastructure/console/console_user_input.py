from simple_agent.application.user_input import UserInput
from simple_agent.application.io import IO
from simple_agent.infrastructure.stdio import StdIO


class ConsoleUserInput(UserInput):

    def __init__(self, indent_level, io: IO):
        self.indent_level = indent_level
        self.io = io or StdIO()

    def read(self) -> str:
        prompt = "\n" + "       " * self.indent_level + "Press Enter to continue or type a message to add: "
        return self.io.input(prompt).strip()

    def escape_requested(self) -> bool:
        return self.io.escape_requested()
