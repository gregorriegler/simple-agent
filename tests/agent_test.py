from approvaltests import verify, Options

from simple_agent.application.input import Input
from simple_agent.application.session import run_session
from simple_agent.infrastructure.console_display import ConsoleDisplay
from simple_agent.infrastructure.console_user_input import ConsoleUserInput
from .print_spy import IOSpy
from .test_helpers import (
    create_temp_file,
    create_temp_directory_structure, all_scrubbers
)
from .test_session_storage import TestSessionStorage
from .test_tool_library import TestToolLibrary


def test_chat_with_regular_response():
    verify_chat(["Hello", "\n"], "Hello! How can I help you?")


def test_chat_with_two_regular_responses():
    verify_chat(["Hello", "User Answer", "\n"], ["Answer 1", "Answer 2"])


def test_chat_with_empty_answer():
    verify_chat(["Hello", ""], "Test answer")


def test_abort():
    verify_chat(["Test message", keyboard_interrupt], "Test answer")


def test_tool_cat(tmp_path):
    temp_file = create_temp_file(tmp_path, "testfile.txt", "Hello world")
    verify_chat(["Test message", "\n"], [f"🛠️ cat {temp_file}", "🛠️ complete-task summary"])


def test_tool_cat_integration(tmp_path):
    temp_file = create_temp_file(tmp_path, "integration_test.txt", "Integration test content\nLine 2")
    verify_chat(["Test message", "\n"], [f"🛠️ cat {temp_file}", "🛠️ complete-task summary"])


def test_tool_ls_integration(tmp_path):
    directory_path, _, _, _, _ = create_temp_directory_structure(tmp_path)
    verify_chat(["Test message", "\n"], [f"🛠️ ls {directory_path}", "🛠️ complete-task summary"])


def test_chat_with_task_completion():
    verify_chat(["Say Hello", "\n"], [
        "Hello!\n🛠️ complete-task I successfully said hello",
        "ignored"
    ])


def test_subagent():
    verify_chat(["Create a subagent that says hello", "\n"], [
        "🛠️ subagent say hello",
        "hello\n🛠️ complete-task I successfully said hello"
    ])


def test_nested_agent_test():
    verify_chat(["Create a subagent that creates another subagent", "\n"], [
        "🛠️ subagent create another subagent",
        "🛠️ subagent say nested hello",
        "nested hello\n🛠️ complete-task I successfully said nested hello",
        "🛠️ complete-task I successfully created another subagent",
        "🛠️ complete-task I successfully created a subagent"
    ])


def test_agent_says_after_subagent():
    verify_chat(["Create a subagent that says hello, then say goodbye", "\n"], [
        "🛠️ subagent say hello",
        "hello\n🛠️ complete-task I successfully said hello",
        "goodbye"
    ])


def test_escape_reads_follow_up_message():
    verify_chat(["Hello", "Follow-up message", "\n"], "Assistant response", [True, False])


def test_escape_aborts_tool_call():
    verify_chat(["Hello", "Follow-up message", "\n"], ["🛠️ cat hello.txt", "🛠️ complete-task summary"], [True, False])


def test_interrupt_reads_follow_up_message():
    verify_chat(["Hello", "Follow-up message", "\n"], "Assistant response", [], [True, False])


def test_interrupt_aborts_tool_call():
    verify_chat(["Hello", "Follow-up message", "\n"], ["🛠️ cat hello.txt", "🛠️ complete-task summary"], [], [True, False])


def verify_chat(inputs, answers, escape_hits=None, ctrl_c_hits=None):
    llm_stub = create_llm_stub(answers)
    message, *remaining_inputs = inputs
    io_spy = IOSpy(remaining_inputs, escape_hits)
    display = ConsoleDisplay(io=io_spy)
    user_input_port = ConsoleUserInput(display.indent_level, io=io_spy)
    user_input = Input(user_input_port)
    user_input.stack(message)
    test_session_storage = TestSessionStorage()
    test_tool_library = TestToolLibrary(llm_stub, io=io_spy, interrupts=[ctrl_c_hits])

    run_session(False, user_input, display, test_session_storage, llm_stub, system_prompt_stub, test_tool_library)

    result = f"# Standard out:\n{io_spy.get_output()}\n\n# Saved messages:\n{test_session_storage.saved}"
    verify(result, options=Options().with_scrubber(all_scrubbers()))


def create_llm_stub(answer):
    if isinstance(answer, list):
        answer_index = 0

        def llm_stub(system_prompt, messages):
            nonlocal answer_index
            if answer_index < len(answer):
                result = answer[answer_index]
                answer_index += 1
                return result
            if answer:
                return answer[-1]
            return ""

        return llm_stub

    def llm_answer(system_prompt, messages):
        return answer

    return llm_answer


def enter(_):
    return "\n"


def keyboard_interrupt(_):
    raise KeyboardInterrupt()


def system_prompt_stub(tool_library):
    return "Test system prompt"
