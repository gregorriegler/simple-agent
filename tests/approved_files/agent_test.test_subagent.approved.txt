# Standard out:
Starting new session

Agent: 🛠️ subagent say hello

       Subagent: hello

       Press Enter to continue or type a message to add: 

       Subagent: 🛠️ complete-task I successfully said hello

       I successfully said hello

# Saved messages:
user: Create a subagent that says hello
assistant: 🛠️ subagent say hello
user: Result of 🛠️ subagent
I successfully said hello
